File: Quick_Sort/Quick_Sort.java
package Quick_Sort;

/**
 * Implementation of the Quick Sort algorithm in Java.
 */
public class Quick_Sort {
    /**
     * Main function to demonstrate Quick Sort on an array.
     *
     * @param args Command line arguments (not used).
     */
    public static void main(String[] args) {
        int[] arr = { 64, 25, 12, 22, 11 };
        System.out.print("Original array: ");
        printArray(arr);
        quickSort(arr, 0, arr.length - 1);

        System.out.print("Sorted array: ");
        printArray(arr);
    }

    /**
     * Recursive function to perform Quick Sort on the array.
     *
     * @param arr  The input array to be sorted.
     * @param low  The starting index of the subarray.
     * @param high The ending index of the subarray.
     */
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Find the partition index
            int partitionIndex = partition(arr, low, high);

            // Recursively sort the subarrays
            quickSort(arr, low, partitionIndex - 1);
            quickSort(arr, partitionIndex + 1, high);
        }
    }

    /**
     * Helper function to find the partition index.
     *
     * @param arr  The input array.
     * @param low  The starting index of the subarray.
     * @param high The ending index of the subarray.
     * @return The partition index.
     */
    public static int partition(int[] arr, int low, int high) {
        // Choose the pivot (last element in this case)
        int pivot = arr[high];
        int i = low - 1;

        // Traverse the array and rearrange elements
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i+1] and arr[high] to place pivot in the correct position
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1; // Return the partition index
    }

    /**
     * Helper function to print an array.
     *
     * @param arr The array to be printed.
     */
    public static void printArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}

File: Quick_Sort/Quick_Sort.cpp
#include <iostream>
using namespace std;

// Function to print elements of an array
void printArray(int arr[], int size)
{
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

// Function to partition the array and return the index of the pivot
int partition(int arr[], int low, int high)
{
    // Choose the last element as the pivot
    int pivot = arr[high];
    int i = low - 1; // Index of the smaller element

    // Traverse through all elements
    for (int j = low; j < high; j++)
    {
        // If the current element is smaller than or equal to the pivot
        if (arr[j] < pivot)
        {
            // Swap arr[i] and arr[j]
            i++;
            swap(arr[i], arr[j]);
        }
    }

    // Swap arr[i+1] and arr[high] to place the pivot in its correct position
    swap(arr[i + 1], arr[high]);
    return i + 1; // Return the index of the pivot
}

// Function to implement the Quick Sort algorithm
void quickSort(int arr[], int low, int high)
{
    if (low < high)
    {
        // Find the partitioning index
        int partitionIndex = partition(arr, low, high);

        // Recursively sort the sub-arrays before and after the partition
        quickSort(arr, low, partitionIndex - 1);
        quickSort(arr, partitionIndex + 1, high);
    }
}

int main()
{
    int arr[] = {64, 25, 12, 22, 11};
    int size = sizeof(arr) / sizeof(arr[0]);

    // Print the original array
    cout << "Original array: ";
    printArray(arr, size);

    // Perform Quick Sort
    quickSort(arr, 0, size - 1);

    // Print the sorted array
    cout << "Sorted array: ";
    printArray(arr, size);

    return 0;
}
